### DISCUSSION TITLE
What needs to be improved in V documentation?

### DISCUSSION THREAD

**Discussion by JalonSolov** [vlang/v#13378] [General]
There has been plenty of discussion on discord about the V documentation, with some ideas mentioned, but no real effort put into making general improvements.

Some ideas that have been mentioned:

- Lower-level "Intro to V".
- Tutorials.
- More details in existing sections.
- Links to the command line help, or at least better descriptions of how to find that help.

So - what are your ideas for improving the V documentation?

**Comment by spytheman** [comment]
Hyperlinking to https://modules.vlang.io/ for the builtin methods, so that people can discover what other builtin methods/functions there are too.

Currently, that is done in only 2 places, that are easy to miss:
```
doc/docs.md:560:[vlib/strconv](https://modules.vlang.io/strconv.html) module.
doc/docs.md:1287:as the `input` function. See the [standard library](https://modules.vlang.io/)
```

**Comment by Hunam6** [comment]
Have a clear and complete table which shows precisely the difference between `struct`s, `union`s and `interface`s.

**Comment by ghost** [comment]
I assume that "V documentation" means https://github.com/vlang/v/blob/master/doc/docs.md, and I'm of the opinion that

A lot of additions have been made since it was first written.
So it would be good to review the sections and reorganize them as necessary.
In the process, would be found that lack of explanation, or things that should be updated.


The first thing that needs to be clarified is the position of V Documentation.

For example, it could be a simple manual for beginners, or a detailed specification and samples, etc.
Depending on this, the structure will be different.

If it is an introductory document, there is no need for a detailed explanation, and it can be links to detailed specification pages, simple examples, links to vlib/examples, etc.

If it's a detailed specification, definition of the syntax, etc.

Will you write all about the V language, or will it be an entry point to more detailed material elsewhere?
If it's the latter, you can post links to specific places. etc.

First of all, it should be clear how it should be positioned.

**Comment by ghost** [comment]
https://github.com/vlang/v/blob/master/cmd/v/help/build.txt
It may be nice to have links to the command line options page.

**Reply by JalonSolov** [reply]
That's in the list in the first post.  :-)

**Comment by spytheman** [comment]
Running this, will show all the public functions/consts/interfaces etc, that have no documentation yet:
`./v run cmd/tools/missdoc.v vlib/ |grep -v .js.v|grep pub`

One metric may be running that each day or on each commit, and monitor whether the number of reported misses grows/shrinks.

**Reply by larpon** [reply]
Small update:
Since #13384 `./v run cmd/tools/missdoc.v vlib/` is now enough

**Comment by spytheman** [comment]
(copied from Discord's help channel, for easier searching/tracking):
I propose a new tool: `v helpus`.

It will have several modes (proposals for others are welcome):
`v helpus document`
=> run the equivalent of the above command (`./v run cmd/tools/missdoc.v vlib/ |grep -v .js.v|grep pub`), then selects a single random function of the undocumented ones (to distribute the work, so that people do not document the same stuff), then whomever is interested, can just run `v helpus document` when he/she has 5 minutes free, do the random task, and make a PR.

`v helpus test`
=> which will get all the public functions, from all modules in vlib, and select one of them at random, and then tell the user to write a test for it and make a PR.

`v helpus confirm`
=> open a browser window with 1 random issue, that is still opened, and not confirmed, and asks the user to check it out, and confirm it with latest V.

**Reply by spytheman** [reply]
it can even choose randomly one of the available modes with just `v helpus`

**Reply by spytheman** [reply]
A simple:
```
Thank you for helping V, 
become the best it can be.
``` 
.. message at the end,
 seems enough to me

**Reply by JalonSolov** [reply]
I agree.

**Reply by spytheman** [reply]
That small sign of politeness will be shown, after the person deliberately typed `v helpus document`. 
I think it is appropriate - the person wanted to help V, and V thanked for that potential help (note that the command will not guarantee that there will be help at all in any direction; it will just guide the person that want to help, to areas, where their efforts will be well spent).

You can argue about the naming - I am opened to actual better suggestions, not pseudo intellectual koans that just waste both your time and my time.

**Reply by spytheman** [reply]
If you do prefer to read bad poetry instead, here is more or less the same, just shorter:
```
Help or not. There is no try.
A small koan riddle, created by Spy.
```

**Reply by JalonSolov** [reply]
Perhaps `help-wanted` (with or without the `-`).  `helpus` isn't bad (with or without a `-`, as in `help-us`).

I like the idea, certainly.  Instead of people asking us what they can help with, or randomly looking through open issues, they can run a simple tool that can suggest areas that need help.

`v helpus` with no options could just print a list of categories where help is needed.
```
$ v helpus
What would you like to help us with?  Try one of the following:

v helpus documentation
v helpus tests
... (etc.)
$
```

**Reply by larpon** [reply]
I also initially though of `v contribute ...` - but it's a bit long (for non-tab-completionists)

**Reply by JalonSolov** [reply]
`contrib` is a fairly common shorter form... 1 char longer than `helpus`, but still may be fine.

**Reply by Hunam6** [reply]
I personally really don't like contrib, I'd largely prefer helpus

**Comment by ZaViBiS** [comment]
I propose to translate docs.md into Ukrainian and Russian.
part of the discussion [here](https://github.com/vlang/v/issues/13410)

**Reply by spytheman** [reply]
I am against that, since it will make updating the main `doc/docs.md` more difficult, and V is still evolving.

**Reply by ZaViBiS** [reply]
Having reread [#13410](https://github.com/vlang/v/issues/13410), I agree that it's too complicated and unreasonable :)

**Reply by spytheman** [reply]
> Having reread [#13410](https://github.com/vlang/v/issues/13410), I agree that it's too complicated and unreasonable :)

Thank you. It is indeed for now - once the syntax is stable, and the feature set fully known, the balance will change.
